modules/mono/csharp_script.h

```diff
@@ -480,6 +490,11 @@ class CSharpLanguage : public ScriptLanguage {
 	virtual String _get_indentation() const;
 	/* TODO? */ void auto_indent_code(String &p_code, int p_from_line, int p_to_line) const override {}
 	/* TODO */ void add_global_constant(const StringName &p_variable, const Variant &p_value) override {}
+	virtual bool has_delayed_script_class_metadata() const override { return true; }
+
+	/* SCRIPT CLASS FUNCTIONS */
+	virtual bool handles_global_class_type(const StringName &p_type) const;
+	virtual StringName get_global_class_name(const String &p_path, StringName *r_base_type = NULL, String *r_icon_path = NULL) const;
```

Comment:
Sorry, I don't get what you're saying here with this reference. Could you please elaborate? Thanks.

Reply:
I initially tried to do it with the global class name stuff, but that alone didn't work. If you change a C# file and in so doing modify its global class name, then the C# file becomes a stale representation of the class information until the code is recompiled. But the engine doesn't dictate *when* the C# class gets recompiled, it's supposed to be driven by the user (e.g. running the compiler or playing a scene which auto compiles beforehand). And that's a separate process than the editorfilesystem gathering up global class names (which it might do quite regularly).

Furthermore, if you open up the editor after closing it, class names have to be loaded in from serialized metadata. They can't pull this metadata from the C# file because the script knows its name only when a compiled build of the script in an assembly exists (and there is none on first bootup), so there's literally no name for the C# file (or possibly a cached old one, iirc). Regardless, I remember that, to solve this issue, I had to create an entirely separate, secondary name caching-mechanism specifically for compiled languages to avoid issues with staleness of data.

Reply:
That makes a lot more sense, thanks for clarifying. The cached data is definitely necessary, and I'm not suggesting you remove it. I'm just wondering if the code that loads that data, something like:
```cpp
if (p_type.type_class->has_attribute(CACHED_CLASS(GlobalAttribute))) {
		MonoObject *attr = p_type.type_class->get_attribute(CACHED_CLASS(GlobalAttribute));
		StringName script_class_name = CACHED_FIELD(GlobalAttribute, name)->get_string_value(attr);
		if (script_class_name != StringName()) {
			r_hint_string = script_class_name;
		}
}
```
could be migrated to `CSharpLanguage::get_global_class_name`, and its current approach of getting the name from the file would only be tried if that failed. I hope that wasn't the possibility you just addressed in your comment, and I've said the exact same thing again.